### DISCUSSION TITLE
Any SSG examples using Next.js + Apollo?

### DISCUSSION THREAD

**Discussion by adamsoffer** [vercel/next.js#10661] [Help] [answered]
I'm interested in trying out the new `unstable_getStaticProps` api using a headless graphql CMS such as Prismic or Contentful. Before I go down this rabbit hole, does anyone have an example of this type of integration they're willing to share?

**Comment by timneutkens** [comment]
Would recommend not using Apollo and just doing a simple `fetch` in getStaticProps. But it works if you just add getStaticProps as getInitialProps in _app is still triggered for incremental adoption purposes.

**Reply by AryanJ-NYC** [reply]
I had great luck migrating away from Apollo Client and using swr and graphql-request.

In case it helps anyone looking here in the future: https://dev.to/aryanjnyc/i-migrated-away-from-apollo-client-to-vercel-swr-and-prisma-graphql-request-and-you-can-too-245b

**Comment by janus-reith** [comment]
How would you deal with usecases where the client needs to take over e.g. to refetch a query, or to use polling? I guess it would work to use something like graphql request first, pass in the data, then in the actual component set up the apollo query in a typical way, probably a hook now,  and ignore that initial data as soon as newly fetched is available, but I was wondering if there is a more seamless approach which does the work once e.g. in the withApollo HOC, while also supporting more edge cases. For example, lets say I have such a specific use case that I dont want to do a separate query on the client, but only have a mutation which would update the cache of existing data that was statically fetched. And then, maybe I even have several of them on separate pages, so on the client I would need to have all necessary data available in the apollo cache.

HOCs wrapping a page will have access to the data which was fetched in getStaticProps of the page, so it would be possible to add to the cache from there. But on the other hand, it would be nice to have a solution where I don't need to specify these queries twice and just have a useQuery somewhere in the Code and that would be fetched on SSG.

**Reply by reachtrevor** [reply]
@janus-reith I was struggling with this as well. I think I've got a clean solution...

The only thing to consider with the below code is that when loading you don't want to return the original state the page loaded with but the state it is in. In this case you could just put an `initial: true` flag inside props inside the return of `getStaticProps` and use that to check further if you want to show the pre-rendered content or whatever was already there.

perhaps @timneutkens can weigh in to see if this is a viable route to pursue.

`withQuery.js`
```js
... imports

const withQuery = (query, options = {}) => {
  return Component => {
    const C = props => {
      const initial = props.data;
      const { data, loading, error } = useQuery(query, options);

      return <Component {...props} data={loading ? initial : data} loading={loading} error={error} />;
    };

    C.displayName = `withQuery(${Component.displayName || Component.name})`;
    C.wrappedComponent = Component;

    return hoistStatics(C, Component);
  };
};
```

`Component.js`
```js
const Component = ({ data }) => {
  return (<div>{data.somekey}<div/>)
};

export async function getStaticProps() {
  return {
    props: {
      data: {
        ... return result of fetch here
      }
    }
  }
}

export default withQuery(GRAPHQL_QUERY_HERE)(Component);
```

**Reply by reachtrevor** [reply]
Simpler, probably more naive version without loading check
`withQuery.js`
```js
const withQuery = (query, options = {}) => {
  return Component => {
    const C = props => {
      const { data, loading, error } = useQuery(query, options);

      return (
        <Component
          {...props}
          data={data || props.data}
          loading={loading}
          error={error}
        />
      );
    };

    C.displayName = `withQuery(${Component.displayName || Component.name})`;
    C.wrappedComponent = Component;

    return hoistStatics(C, Component);
  };
};
```

**Reply by janus-reith** [reply]
@trevordmiller Thanks, yes basically I already ended up doing something similar, although in a hook and not a hoc. I realized that usually there was no need to have that initial data present as part of the apollo cache, so I could keep it just that simple aswell.

**Reply by adamsoffer** [reply]
@trevthewebdev Interesting thanks for sharing. Is there an example of this anywhere? Would be great to get a version of the `withApollo` example using this SSG technique.

**Reply by reachtrevor** [reply]
@adamsoffer 
My second iteration is working well, no more weird loading states and no exceptions because I can return `data['someKey']` if for some reason the query fails.

I have not tried it using `withApollo` honestly I have a lot of experience with react but I am a beginner with graphQL and apollo. My use case was I wanted statically generated pages backed by strapi (headless cms) the above implementation seemed to accomplish my goals for the time.

**Reply by adamsoffer** [reply]
@janus-reith mind sharing your hook example? Trying to do something similar myself.

**Reply by janus-reith** [reply]
@adamsoffer The example @nickluger linked where the initial cache is passed to Apollo seems to be a clean solution, provided that the overhead of using Apollo on the server and extracting the cache is not too much of an overhead compared to something simpler on the server like `graphql-request`.

My hook doesn't really do anything special. You fetched the `initialData`, and have it available as a prop on your page function.
There you have something like this where you pass initialData to the hook:
`const [initialDataOrFromApollo, refetch, isLoading] = useInitalDataOrApolloData(initialData);`

In that hook you have `useQuery` somewhere and at the end return the apollo query data or just pass through the initial data the hook recieved:
```
return [
	dataFromUseQuery || initialData,
	refetch
	isLoading
]
```

So just as the solution @trevthewebdev provided, this doesn't actually take care of having initial data available in apollo, but passing similar data to your components as a substitute.

**Comment by nickluger** [comment, marked as answer]
Heres a small example: https://github.com/zeit/next.js/discussions/11957#discussioncomment-7190

**Reply by janus-reith** [reply]
I would consider this example the actual answer to this, as it covers the idea of having Apollo with initial data from `getStaticProps` available.

**Reply by adamsoffer** [reply]
ah sweet yeah that's what I'm looking for :)
